car-bridge/src/main.rs

```diff
@@ -29,58 +41,193 @@ async fn main() -> Result<(), Box<dyn Error>> {
-                        info!("Trying to reconnect: {}.", err);
-                        sleep(Duration::from_secs(5)).await;
-                    }
-                }
+                let mut chariott = chariott.clone();
+                let response_sender = response_sender.clone();
+
+                spawn(async move {
+                    handle_message(&mut chariott, response_sender, message).await;
+                });
             }
             _ = cancellation_token.cancelled() => {
+                debug!("Shutting down the subscriber loop.");
                 break;
             }
         }
     }
 
-    info!("Disconnecting the client.");
-    client.disconnect(None).await?;
+    publish_handle.await?;
 
     Ok(())
 }
+
+async fn handle_message(
+    chariott: &mut impl ChariottCommunication,
+    response_sender: Sender<(String, MessageBuilder)>,
+    message: MqttMessage,
+) {
+    let correlation_information = match message.get_correlation_information() {
+        Ok(cm) => cm,
+        Err(error) => {
+            debug!("Error when getting correlation information from message: '{error:?}'.");
+            return;
+        }
+    };
+
+    let response: Result<_, Box<dyn std::error::Error + Send + Sync>> = async {
+        let fulfill_request: FulfillRequest = Message::decode(message.payload())?;
+
+        let intent_enum = fulfill_request
+            .intent
+            .and_then(|intent| intent.intent)
+            .ok_or_else(|| Error::new("Message does not contain an intent."))?;
+
+        let response = match intent_enum {
+            IntentEnum::Discover(_) => Err(Error::new("Discover is not supported.")),
+            IntentEnum::Subscribe(_) => todo!(),
+            _ => chariott
+                .fulfill(fulfill_request.namespace, intent_enum)
+                .await
+                .map(|r| r.into_inner()),
+        }?;
+
+        let mut payload = vec![];
+        response.encode(&mut payload)?;
+
+        Ok(Response {
+            payload,
+            content_type: "application/x-proto+chariott.common.v1.FulfillResponse",
+            is_error: false,
+        })
+    }
+    .await;
+
+    let response = match response {
+        Ok(message) => message,
+        Err(error) => {
+            debug!("Error when handling message: '{error:?}'.");
+
+            let message = ValueMessage { value: Some(ValueEnum::String(format!("{error:?}"))) };
+
+            let mut payload = vec![];
+            if let Err(err) = message.encode(&mut payload) {
+                debug!("Failed to encode error response: '{err}'.");
+            }
+
+            Response {
+                payload,
+                content_type: "application/x-proto+chariott.common.v1.Value",
+                is_error: true,
+            }
+        }
+    };
+
+    let mut properties = Properties::new();
+
+    properties.push_string(PropertyCode::ContentType, response.content_type).unwrap();
```

Comment:
We're unwrapping/panicking for this and setting of the error property, but not for the correlation data property. To avoid this inconsistency, we should either panic on failure or trace and exit when setting any property.

Reply:
The difference that I wanted to express with the panic/error is that both the content type and the error user property are under the Car Bridge's control - if we suddenly fail to set those, it's a bug.

The correlation data is driven by the publisher, and hence it is not under the Car Bridge's control, which is why I was failing differently.

I used recoverable errors for consistency with 06418b39e423783d79a9e1eda3cfee15975a62e7, which also introduces some duplication in how we handle these errors. I would suggest to not attempt to fix this, at it will look differently in the follow-up PR #63.